### PR TITLE
MCR-2202 automatic pi creation

### PR DESCRIPTION
**Pull request by sebhofmann**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2202).

### CODE REVIEW

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 7
           

Complexity increasing per file
==============================
- mycore-pi/src/main/java/org/mycore/pi/MCRPICreationEventHandler.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIOrPredicate.java  1
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIXPathPredicateTest.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPICombinedPredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIPublishedPredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIPredicateBase.java  3
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRTruePredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIXPathPredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIStatePredicate.java  1
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIAndPredicate.java  1
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRFalsePredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/MCRPIJobService.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577&bid=17936188)

File: mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIPredicateBase.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.mycore.common.config.MCRConfiguration2;
+import org.mycore.common.config.MCRConfigurationException;
+
+public abstract class MCRPIPredicateBase implements MCRPIPredicate {
+
+    public MCRPIPredicateBase(String propertyPrefix){
+        this.propertyPrefix = propertyPrefix;
+    }
+
+    private final String propertyPrefix;
```
Comment (adlerre) [inline mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIPredicateBase.java:14]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577)

File: mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        final boolean mock2Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_2 + ".CreationPredicate").test(emptyBase);
+        final boolean mock3Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_3 + ".CreationPredicate").test(emptyBase);
+        final boolean mock4Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_4 + ".CreationPredicate").test(emptyBase);
+
+        Assert.assertTrue(mock1Result);
+        Assert.assertFalse(mock2Result);
+        Assert.assertTrue(mock3Result);
```
Comment (adlerre) [inline mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java:38]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577)

File: mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_1 + ".CreationPredicate").test(emptyBase);
+        final boolean mock2Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_2 + ".CreationPredicate").test(emptyBase);
+        final boolean mock3Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_3 + ".CreationPredicate").test(emptyBase);
+        final boolean mock4Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_4 + ".CreationPredicate").test(emptyBase);
+
+        Assert.assertTrue(mock1Result);
+        Assert.assertFalse(mock2Result);
```
Comment (adlerre) [inline mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java:37]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertFalse' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577)

File: mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIXPathPredicate.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+public class MCRPIXPathPredicate extends MCRPIPredicateBase
+    implements MCRPICreationPredicate, MCRPIObjectRegistrationPredicate {
+    public MCRPIXPathPredicate(String propertyPrefix) {
+        super(propertyPrefix);
+        final String xPath = requireProperty("XPath");
+        XPathFactory factory = XPathFactory.instance();
+        expr = factory.compile(xPath, Filters.fpassthrough(), null, MCRConstants.getStandardNamespaces());
+    }
+
+    final private XPathExpression<Object> expr;
```
Comment (adlerre) [inline mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIXPathPredicate.java:18]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577)

File: mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        final boolean mock1Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_1 + ".CreationPredicate").test(emptyBase);
+        final boolean mock2Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_2 + ".CreationPredicate").test(emptyBase);
+        final boolean mock3Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_3 + ".CreationPredicate").test(emptyBase);
+        final boolean mock4Result = MCRPIJobService
+            .getPredicateInstance("MCR.PI.Service." + MOCK_SERVICE_4 + ".CreationPredicate").test(emptyBase);
+
+        Assert.assertTrue(mock1Result);
```
Comment (adlerre) [inline mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java:36]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertTrue' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577)

File: mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        Assert.assertFalse(mock2Result);
+        Assert.assertTrue(mock3Result);
+        Assert.assertFalse(mock4Result);
```
Comment (adlerre) [inline mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java:39]: Codacy Issue found: [Unnecessary use of fully qualified name 'Assert.assertFalse' due to existing static import 'org.junit.Assert.*'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577)

File: mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.Map;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.mycore.common.MCRTestCase;
+import org.mycore.datamodel.metadata.MCRBase;
+import org.mycore.pi.MCRPIJobService;
+
+import static org.junit.Assert.*;
```
Comment (adlerre) [inline mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java:11]: Codacy Issue found: [Avoid unused imports such as 'org.junit.Assert'](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577)

File: mycore-pi/src/main/java/org/mycore/pi/MCRPIJobService.java [hunk 8]
```diff
@@ -325,43 +327,61 @@ protected PiJobAction getAction(Map<String, String> contextParameters) {
...
+    protected Predicate<MCRBase> getRegistrationCondition() {
+        final String predicateProperty = MCRPIServiceManager.REGISTRATION_SERVICE_CONFIG_PREFIX +
+            getServiceID() + "." + MCRPIJobService.REGISTRATION_PREDICATE;
+        if(MCRConfiguration2.getString(predicateProperty).isEmpty()){
+            return (o)->true;
+        }
+        return getPredicateInstance(predicateProperty);
+    }
 
+    public static Predicate<MCRBase> getPredicateInstance(String predicateProperty) {
```
Comment (rsteph-de) [inline mycore-pi/src/main/java/org/mycore/pi/MCRPIJobService.java:348]: Is there no utilitiy method in MCRConfiguration2 for instantiating classes?

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIXPathPredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIPredicateBase.java  3
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIAndPredicateTest.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIPublishedPredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIOrPredicate.java  1
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRTruePredicate.java  1
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRFalsePredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIStatePredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPICombinedPredicate.java  1
- mycore-pi/src/main/java/org/mycore/pi/condition/MCRPIAndPredicate.java  1
- mycore-pi/src/test/java/org/mycore/pi/condition/MCRPIXPathPredicateTest.java  1
- mycore-pi/src/main/java/org/mycore/pi/MCRPICreationEventHandler.java  1
- mycore-pi/src/main/java/org/mycore/pi/MCRPIJobService.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5472577&bid=17936188)

**Review by rsteph-de — APPROVED** [review]
looks good for me

**Pull request merged**